Add Azurestack Support

patrickdillon opened the pull request:
Updated as this is no longer WIP.

This PR:

- adds a bare-bones azurestack provider
- copies all of the logic from the azure provider
- adjusts the logic behind hostname fetching to query the identity metadata service (not the new instance metadata endpoint) 
- removes support for attributes (using the logic of azure resulted in incorrect IP addresses)

I tested this and it pulled unique hostnames for a bootstrap node and 3 masters. The SSH keys also appear in the login screen. 

Note: because this uses the identity metadata service, this will work in any azure stack environment.

Further todo on this PR:

- [X] remove changes to dracut/30afterburn/afterburn-hostname.service
- [X] Go through all of the logic from the azure provider and determine whether it is necessary; strip it out if we don't need it
- [X] Determine whether we want to query both the compute endpoint and the identity endpoint, falling back to one if the other doesn't work
- [X] Clean up commits (better messages, squash FIXUPs)

Regarding the compute endpoint: 
From a running VM (in the lab environment) I am able to query it:
```shell
[core@ash-bootstrap ~]$ curl -H Metadata:true --noproxy "*" http://169.254.169.254/metadata/instance?api-version=2019-03-11 | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1484  100  1484    0     0   9160      0 --:--:-- --:--:-- --:--:--  9217
{
  "compute": {
    "azEnvironment": "AzureStack",
```
But when I hit that same endpoint during afterburn I get an error. Because I was able to get the vm name from the identity endpoint, I prioritized that. as future work we could add more support for the instance metadata service.

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -1,14 +1,9 @@
-ConditionKernelCommandLine=|ignition.platform.id=exoscale
-ConditionKernelCommandLine=|ignition.platform.id=ibmcloud
-ConditionKernelCommandLine=|ignition.platform.id=vultr
+ConditionKernelCommandLine=|ignition.platform.id=azurestack
```
> there is a messy change in afterburn-hostname.service that breaks other platforms, but I think I should be able to just add them back in....

Yeah, there shouldn't be any issue there at all with just adding this line only to the original service unit.

jlebon commented on src/providers/azurestack/crypto/mod.rs:
```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 CoreOS, Inc.
```
`2021`

This came up in discussion in another project as well. For this project at least, we should probably be able to drop all these blocks and just rely on the `Cargo.toml` license field. (To be clear, I mean this as a follow-up, not something you need to do here.)

jlebon commented on src/providers/azurestack/mod.rs:
```diff
@@ -0,0 +1,369 @@
+
+    fn fetch_hostname(&self) -> Result<Option<String>> {
+        let instance_metadata = AzureStack::fetch_identity();
+        Ok(Some(instance_metadata.unwrap().vm_name))
```
The linter is indirectly saying to use `Err` instead of `unwrap()`, e.g.:

```suggestion
        let instance_metadata = AzureStack::fetch_identity()
        	.context("fetching instance identity")?;
        Ok(Some(instance_metadata.vm_name))
```

patrickdillon replied:
Is there a way I should change `fetch_identity`: https://github.com/coreos/afterburn/pull/561/files#diff-34cbce01c9aa769653cb8f06a598911f28d90de5710e9b602560660669d702d1R172-R183 where I already have `context`?

jlebon replied:
The `fetch_identity` function looks good. The `context` stuff is like a human-friendly backtrace where you can add more error-prefixing as the error bubbles up the stack to make it easier to tell where the error came from. It's up to you when to add a `.context(...)?` vs just `?`. If you think e.g. it's repetitive with the context already added by the callee, you can omit it in the caller. In this case for example, yeah it's probably superfluous given that my suggestion is more or less repeating the context provided by `fetch_identity`, and our caller ultimately adds the context that the failure happened while fetching the hostname:

```rust
        // write hostname if configured to do so
        self.hostname_file
            .map_or(Ok(()), |x| metadata.write_hostname(x))
            .context("writing hostname")?;
```

So you can probably just swap out `.context("fetching instance identity")?` with just `?` in my suggestion above.

jlebon commented on src/providers/azurestack/mod.rs:
```diff
@@ -0,0 +1,369 @@
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        Ok(HashMap::new())
+    }
+
```
That's the default implementation, so you can drop it entirely.

jlebon commented on src/providers/azurestack/mod.rs:
```diff
@@ -0,0 +1,369 @@
+    pub vm_name: String,
+    #[serde(rename = "subscriptionId", default)]
+    pub subscription_id: String,
+}
```
Should be able to simplify to

```suggestion
#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
struct InstanceMetadata {
    pub vm_name: String,
    pub subscription_id: String,
}
```

Do we actually need the `default` here? I.e. do we always expect those fields to exist, or do we want to support them not existing? In case of the former, we should just drop `default`. In case of the latter, it would be cleaner to use an `Option<String>` instead.

jlebon committed:
providers: move azure into microsoft module

Prep for adding an `azurestack` provider which shares a lot with
`azure`.

patrickdillon committed:
providers: add Azure Stack Hub (azurestack)

Adds support and tests for the Azure Stack platform. Sets the hostname
based on the vm name. Obtains SSH keys. Checks in to finish the boot
process. Unlike Azure, does not offer any attributes. Follows the
same pattern as Azure with some notable differences:

The hostname is obtained by parsing the vm name from the identity
metadata service, similar to the WALinuxAgent. Early attempts were
made to obtain this vm name from the compute instance metadata service
which was available in a pre-release lab environment, but the endpoint
was not reachable pre-check in. Consdering the vm name was available
from the identity endpoint, accessing the instance metadata service
can be saved for future work.

SSH key functionality is the same as Azure.

Regarding attributes, obtaining the IP addresses in the same manner as
Azure led to incorrect values, so they were removed as well as vmsize
for simplicity. Therefore no attributes are available.

Check in is the same process as Azure.

Co-authored-by: Ben Howard <ben.howard@redhat.com>
Co-authored-by: Jonathan Lebon <jonathan@jlebon.com>

jlebon reviewed (approved):
I've also finally been able to successfully test this in ASH, including SSH key fetching (which matters more for FCOS):

```
[root@jlebon-vm-bootstrap ~]# journalctl -u afterburn-sshkeys@core.service
-- Logs begin at Thu 2021-04-08 01:43:58 UTC, end at Thu 2021-04-08 03:27:48 UTC. --
Apr 08 02:00:22 jlebon-vm-bootstrap systemd[1]: Starting Afterburn (SSH Keys)...
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.046 WARN Failed to get fabric address from DHCP: maximum number of retries (60) reached
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.046 INFO Using fallback address
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.055 INFO Fetching http://168.63.129.16/?comp=versions: Attempt #1
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.062 INFO Fetch successful
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.062 INFO Fetching http://168.63.129.16/machine/?comp=goalstate: Attempt #1
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.219 INFO Fetch successful
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.270 INFO Fetching http://168.63.129.16/machine/70f676d8-fb84-4921-9716-96cc629c05a7/...
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: Apr 08 02:00:51.577 INFO Fetch successful
Apr 08 02:00:51 jlebon-vm-bootstrap afterburn[1929]: wrote ssh authorized keys file for user: core
Apr 08 02:00:51 jlebon-vm-bootstrap systemd[1]: afterburn-sshkeys@core.service: Succeeded.
Apr 08 02:00:51 jlebon-vm-bootstrap systemd[1]: Started Afterburn (SSH Keys).
[root@jlebon-vm-bootstrap ~]# cat /home/core/.ssh/authorized_keys.d/afterburn
ssh-rsa AAA...
```

So my vote is to get this in and we can do more cleanups afterwards. @lucab Would appreciate a stamp to sanity-check there isn't anything majorly wrong.
